### PR TITLE
Fix detokenization of non-special added-tokens

### PR DESCRIPTION
**Pull request by goerch**
Checked with

```
.\build\bin\Release\main.exe -m models\mpt-7B-storywriter\ggml-model-f16.gguf -p "Once upon a time  there" --temp 0 -n 32
```

Before:

```
Once upon a timethere
```

After:

```
Once upon a time  there
```

Also fixes build problems with `clang` on `Windows` (not detected by CI?). We can't link with `clip` and `common` here due to duplicate symbol errors.

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -9788,12 +9795,15 @@ int llama_token_to_piece(const struct llama_model * model, llama_token token, ch
...
                 memcpy(buf, result.c_str(), result.length());
                 return result.length();
+            } else if (llama_is_user_defined_token(model->vocab, token)) {
+                std::string result = model->vocab.id_to_token[token].text;
+                if (length < (int) result.length()) {
+                    return -result.length();
+                }
+                memcpy(buf, result.c_str(), result.length());
+                return result.length();
             } else if (llama_is_control_token(model->vocab, token)) {
```
Comment (Galunid) [inline llama.cpp:9805]: Seems redundant, since we ignore the token regardless of it's a control one or not.

Comment (goerch) [reply llama.cpp:9805]: Yes. But these are the token types I'd expect to be used. In an older version the code asserted that no other token type could occur, but I'm not sure how we should react (assertion, exception or ignorance).

**Comment by Galunid** [comment]
I tested it with #3586 and it works when passing `  ` (replaced with `__`) as part of the prompt, thanks!
> The best music is__the kind that makes you feel as though it was written for you, ad nothing more. There's a special way to listen to this music: the way a friend might play an old recording of their favourite band, over and over again, until the grooves are worn down to nubs. Or the way someone might tell a story they've never told anyone else about how they met that

**Comment by goerch** [comment]
When checking

```
    std::vector<llama_token> tokens = llama_tokenize(ctx, "[PAD50277]", false, true);
```

with `mpt` this resolves to token 50277. This looks wrong to me, should I try to fix this in this PR, too?

**Comment by goerch** [comment]
Referencing a comment with [my understanding](https://github.com/ggerganov/llama.cpp/pull/3586#issuecomment-1777687265) of the situation I see a couple of options:

- more compatibility with HF tokenizers (might be reasonably easy for our tokenizers, but with some risk breaking the remainder of the stack?)
- do nothing (token classification might be questionable for added special tokens and padding tokens)
- fix the problems(bug?) with added special tokens being classified as USER_DEFINED (some risk breaking the remainder of the stack)
- fix the problem with padding tokens being classified as USER_DEFINED (what should the token type be: UNUSED?)
     
I'll need some guidance here on how to best proceed.

**Comment by cebtenzzre** [comment]
Weren't the padding tokens originally a hack anyway? I don't know that they should be included in the actual tokenizer vocabulary. They only exist because GGUF doesn't (currently) have a way to specify vocab_size.
https://github.com/ggerganov/llama.cpp/blob/ad939626577cd25b462e8026cc543efb71528472/convert-gptneox-hf-to-gguf.py#L121-L122

**Comment by cmp-nct** [comment]
> Weren't the padding tokens originally a hack anyway? I don't know that they should be included in the actual tokenizer vocabulary. They only exist because GGUF doesn't (currently) have a way to specify vocab_size.
> 
> https://github.com/ggerganov/llama.cpp/blob/ad939626577cd25b462e8026cc543efb71528472/convert-gptneox-hf-to-gguf.py#L121-L122

I don't think that's the case, though not sure about the origin of it.

Quite often the python HF models are released with tensors and a vocabulary sized larger than the actually real vocabulary is. OpenAssistant (Falcon) for example has 12 missing tokens compared to the tensor size and to the actual vocab_size specified.

I believe the correct way is to generate PAD tokens, they can not be generated anyway it's a placeholder. 
Originally it's probably also a placeholder for continued fine-tuning that will unlikely ever be utilized.

**Comment by cebtenzzre** [comment]
> I believe the correct way is to generate PAD tokens

As far as I understand, HF transformers does not need these PAD tokens in the vocabulary, so why not make things simple and try to be as close to their behavior as we can? We are trying to get the tokenizer to ignore these tokens when they don't really exist in the first place. We could use the UNUSED token type, but it seems like unnecessary effort to generate placeholders just so they can be totally ignored by the implementation.

**Comment by TheBloke** [comment]
Yes, it'd be awesome if there was a fix that meant these extra tokens weren't needed.  Recently CausalLM released with a whopping 213 missing tokens, comparing config.vocab_size to supplied vocabulary.

My 'fix' for that until now has been to add a bunch of `<dummyXXX>`tokens to added_tokens.json. Kerfluffle recently provided an official implementation of that via --padvocab which I appreciate, though it's obviously still not a very elegant solution overall.

I agree with @cebtenzzre that it'd be really great if llama.cpp could just ignore these 'phantom' tokens, like transformers does. 

As to why models do this - in some cases it's because they're rounding the vocab size up to a multiple so that tensor parallelism works. Eg models that have a vocab size of 32032 or 32128 - their actual vocab might be, say, 32003, but that breaks TP so they then round up to the next nearest multiple.   Though that's not why OpenAsisstant did it.

**Comment by cmp-nct** [comment]
You are right, ignoring the mismatch is a better way to solve it. But I'd guess the stored vocab size in converted model should actually reflect the true vocabulary size and not continue with the wrong one.

Basically it's misbehavior on config side to claim a false vocabulary size, most likely to satisfy something in the transformers libs which then in sequence ignore the wrong number.

Adding PAD tokens is/was a quick fix for a wrong config, so either adding PAD tokens or correcting the wrong size number and relying on tensor size for calculations. It probably needs a fix in all eval routines and checks that rely on vocab size.

**Comment by ggerganov** [comment]
Continued here: https://github.com/ggerganov/llama.cpp/pull/3760

**Pull request closed without merging**